subprojects/snapshots/src/test/groovy/org/gradle/internal/snapshot/AbstractSnapshotWithChildrenTest.groovy

```diff
@@ -0,0 +1,208 @@
+    }
+
+    String getPathFromCommonPrefix() {
+        return absolutePath.substring(offset + commonPrefix.length() + 1)
+    }
+
+    String getSelectedChildPathFromCommonPrefix() {
+        return selectedChild.pathToParent.substring(commonPrefix.length() + 1)
+    }
+
+    def getDescendantSnapshotOfSelectedChild(@Nullable MetadataSnapshot foundSnapshot) {
+        def descendantOffset = offset + selectedChild.pathToParent.length() + 1
+        1 * selectedChild.getSnapshot(absolutePath, descendantOffset, CASE_SENSITIVE) >> Optional.ofNullable(foundSnapshot)
+    }
+
+    def invalidateDescendantOfSelectedChild(@Nullable FileSystemNode invalidatedChild) {
+        def descendantOffset = offset + selectedChild.pathToParent.length() + 1
+        1 * selectedChild.invalidate(absolutePath, descendantOffset, CASE_SENSITIVE) >> Optional.ofNullable(invalidatedChild)
+    }
+
+    @SuppressWarnings("GrMethodMayBeStatic")
+    void noMoreInteractions() {
+        _ * _.pathToParent
+        0 * _
+    }
+
+    /**
+     * Different lists of relative paths of the initial children of the node under test.
+     */
+    static final INITIAL_CHILD_CONSTELLATIONS = [
+        ['name'],
+        ['name', 'name1'],
+        ['name', 'name1', 'name12'],
+        ['name', 'name1', 'name12', 'name2', 'name21'],
+        ['name/some'],
+        ['name/some/other'],
+        ['name/some', 'name2'],
+        ['name/some', 'name2/other'],
+        ['name/some', 'name2/other'],
+        ['name', 'name1/some', 'name2/other/third'],
+        ['aa/b1', 'ab/a1', 'name', 'name1/some', 'name2/other/third'],
+        ("a".."z").toList(),
+    ]
+
+    /**
+     * The queried/updated path has no common prefix with any of the initial children of the node under test.
+     *
+     * E.g.
+     *   path: 'name0/some
+     *   children: ['name/some', 'other']
+     * or
+     *   path: 'completelyDifferent'
+     *   children: ['name/some', 'other']
+     */
+    static final List<VirtualFileSystemTestSpec> NO_COMMON_PREFIX = INITIAL_CHILD_CONSTELLATIONS.collectMany { childPaths ->
+        childPaths.collect { childPath ->
+            def firstSlash = childPath.indexOf('/')
+            String newChildPath = firstSlash > -1
+                ? "${childPath.substring(0, firstSlash)}0${childPath.substring(firstSlash)}"
+                : "${childPath}0"
+            new VirtualFileSystemTestSpec(childPaths, newChildPath, 0, null)
+        } + new VirtualFileSystemTestSpec(childPaths, 'completelyDifferent', 0, null)
+    } + new VirtualFileSystemTestSpec([], 'different', 0, null)
+
+    /**
+     * The queried/updated path has a true common prefix with one of the initial children of the node under test.
+     *
+     * E.g.
+     *   path: 'name/different'
+     *   children: ['name/some', 'other']
+     */
+    static final List<VirtualFileSystemTestSpec> COMMON_PREFIX = INITIAL_CHILD_CONSTELLATIONS.collectMany { childPaths ->
```

Comment:
And here we are manipulating a sibling(?) of a grandchild?

Reply:
No. The selected path is a sibling of the children, only there is a common prefix between the selected path and one of the children. That means when adding the new child, we need to insert an intermediate node between the current node and the selected child.